[WIP] added reg.imported state module function

H20-17 opened the pull request:
### What does this PR do?
Adds the state function, _reg.imported_file_, which is supposed to be the stateful correlate of the existing executable module function, _reg.import_file_

### What issues does this PR fix or reference?
This does not fix an existing PR (AFAIK)

### Tests written?

No

### Commits signed with GPG?

No

Commit:
added reg.imported_file state module function

Commit:
Some refactoring to satisfy Codeclimate complaints

Commit:
Some refactoring to satisfy Codeclimate complaints

Commit:
Some refactoring to satisfy Codeclimate complaints

Commit:
Some refactoring to satisfy Codeclimate complaints

Commit:
Some refactoring to satisfy Codeclimate complaints

Commit:
Some refactoring to satisfy Codeclimate complaints

Commit:
Some refactoring to satisfy Codeclimate complaints

Commit:
Some refactoring to satisfy Codeclimate complaints

Commit:
Lint fix

Commit:
Lint fix

Commit:
Lint fix

H20-17 commented:
This is bug free and does exactly what I think it should but I decided that I could improve the code a lot, so I'm refactoring.

cmcmarrow commented:
Hey H20-17! Thanks for the code. It looks good. Can you please write some test for your code? If you need help with writing the test please let me know. I'll be happy to help you.

H20-17 commented:
> Hey H20-17! Thanks for the code. It looks good. Can you please write some test for your code? If you need help with writing the test please let me know. I'll be happy to help you.

Hi @cmcmarrow, I think that the code is good insofar as it works (i.e. you get the right output for the input you give it), but I think that it's not very elegant and doesn't capture logic of salt states in the most concice way. It's still WIP.

Yes, I confess i know nothing about testing. I may need assistance with that. As soon as this is non-WIP (which should be appxox. tonight or tomorrow some time as soon as I get the time to make the changes I want), I'd like some pointers about what I'm supposed to do about that because I haven't got a clue.

H20-17 commented:
Please no one review the code until this ceases to be WIP. I'm happy to have people test it by running it since its working and the functionality is complete, and the functionality will not change because of the refactoring.

Commit:
Lint fix

H20-17 commented:
This is no longer WIP. The github lint test is not completing. The function is now just called "imported" instead of "imported_file. The code has been refactored a lot to be more clear.

Commit:
Add missing return value

Commit:
Fix comment typo

damon-atkins commented on salt/states/reg.py:
```diff
@@ -617,3 +621,317 @@ def key_absent(name, use_32bit_registry=False):
+        # The first line of a reg file is english text which we must consume before parsing.
+        # It contains no data.
+        reg_data = configparser.ConfigParser()
+        PY2 = sys.version_info[0] == 2
```
PY2 is already defined see other salt code for examples.

damon-atkins commented on salt/states/reg.py:
```diff
@@ -617,3 +621,317 @@ def key_absent(name, use_32bit_registry=False):
+        word_sz_txt = "64"
+    present_reg_file = (__salt__['temp.file'])()
+    try:
+        cmd = 'reg export "{0}" "{1}" /y /reg:{2}'.format(reg_location, present_reg_file, word_sz_txt)
```
This assumes all items are under a single tree.

damon-atkins commented on salt/states/reg.py:
```diff
@@ -617,3 +621,317 @@ def key_absent(name, use_32bit_registry=False):
+    changes = {}
+    operation = None
+    if not _imported_test(state_parse, reference_data):
+        changes['old'] = "Registry unmodified."
```
These should be empty.     The `comment `filed is the only field to have comments.  new and old should only contain what the old values & delete items and new contain changes to existing values & additional items.

H20-17 replied:
> These should be empty. The `comment `filed is the only field to have comments. new and old should only contain what the old values & delete items and new contain changes to existing values & additional items.

The changes dictionary i'm returning gets used when in test mode or later on if the operation is performed and succeeds (I hang on to to is so I can use it in the final output), so it needs to not be empty in the situations where i'm actually using it.

I realise that the contents of the dictionary isn't standards conformant. I can fix that but it'll be a few minutes of work since I really should be creating the changes dictionary as I'm testing. It's not a big deal.

damon-atkins reviewed (changes requested):
Whats the aim of the code.  I think I known what you are trying to do.  Bit of introduction is needed.

What actions are taken if something did exist in the file and and later in the future it no longer exists.
What are the limitations,  e.g. everything needs to be under the same tree?
Are you better off not using the reg.exe at all, given you need to parse the file anyway?

damon-atkins commented:
PS   [WIP] at the front of the description to indicate it is work in progress

H20-17 commented:
> Whats the aim of the code. I think I known what you are trying to do. Bit of introduction is needed.
> 
> What actions are taken if something did exist in the file and and later in the future it no longer exists.
I don't understand this question.
> What are the limitations, e.g. everything needs to be under the same tree?
Yes it assumes that the reg file is the consequence of a reg export, which I think implies a single tree. Under what situations might we end up with other types of reg files? I'll clarify that in the text if you think this is an issue.
> Are you better off not using the reg.exe at all, given you need to parse the file anyway?
It keeps the code simpler because I don't have to interpret the meaning of values in reg files. Instead I,m just do a string comparison of parsed values without worrying about types. Also this is _possibly_ faster. If I don't grab a section of the registry and turn it into some kind of data, the alternative is a lot of individual lookups. It sounds slow-ish but perhaps it isn't.

Commit:
Fix wrong return values

Commit:
Clarification

Commit:
Refactoring + PY2 issue addressed + changes dict made more standards conformant + limitations noted in documentation

H20-17 commented:
@damon-atkins I have attempted to address all your issues now apart from the multiple tree roots issue (but i mentioned this a limitation in the documentation) and I am still just exporting and comparing the parse trees which I sitll think is the right approach. I have made sure that the return changes dictionary now has good data. 

The changes dictionary ends up looking like this:
```
              new:
                  ----------
                  added keys:
                      - HKEY_LOCAL_MACHINE\SOFTWARE\SaltTestKey\subKey5
                  added/changed values:
                      ----------
                      HKEY_LOCAL_MACHINE\SOFTWARE\SaltTestKey\subKey3:
                          |_
                            ----------
                            @:
                                oik
                          |_
                            ----------
                            g:
                                ciuh
                      HKEY_LOCAL_MACHINE\SOFTWARE\SaltTestKey\subKey5:
                          |_
                            ----------
                            @:
                                oki2
                          |_
                            ----------
                            g:
                                c
                          |_
                            ----------
                            xterm256colour:
                                dword:10001011
              old:
                  ----------
                  values:
                      ----------
                      HKEY_LOCAL_MACHINE\SOFTWARE\SaltTestKey\subKey3:
                          |_
                            ----------
                            @:
                                ok
                          |_
                            ----------
                            g:
                                ch

```

sagetherage commented:
@H20-17 please rebase this against the master branch

sagetherage commented:
@H20-17 please rebase this against the master branch

sagetherage commented:
Reviewed in Grooming 2021-May-03 considering closing
